### PR TITLE
[Bugfix] Fix missing `post_layernorm` in CLIP

### PR DESCRIPTION
**Pull request by DarkLight1337**
CLIP encoder currently assumes that `len(self.encoder.layers) < config.num_hidden_layers` which may not always be the case.

This is a follow-up to [#8106](https://github.com/vllm-project/vllm/pull/8106#issuecomment-2326268446). 

~It additionally fixes a potential issue in SigLIP where the weights are not sharded like in CLIP. cc @Isotr0py~ Fixed in #8269

### CODE REVIEW

**Comment by Isotr0py** [comment]
The modification of siglip loading logic looks good to me.

BTW, are the disabled `"TIGER-Lab/Mantis-8B-siglip-llama3"` model test also related to this `post_layernorm` issue? I saw it was disabled before.
https://github.com/vllm-project/vllm/blob/d3311562fbe740a883e7f03f0b59620587cabb29/tests/models/test_llava.py#L23-L27

**Comment by DarkLight1337** [comment]
> The modification of siglip loading logic looks good to me.
> 
> BTW, are the disabled `"TIGER-Lab/Mantis-8B-siglip-llama3"` model test also related to this `post_layernorm` issue? I saw it was disabled before.
> 
> https://github.com/vllm-project/vllm/blob/d3311562fbe740a883e7f03f0b59620587cabb29/tests/models/test_llava.py#L23-L27

Yes, that's how I discovered it. However even after fixing the weight loading, the model cannot be run because of the special preprocessing they use.

File: vllm/model_executor/models/siglip.py [hunk 1]
```diff
@@ -443,27 +443,19 @@ def __init__(
         self.config = config
         embed_dim = config.hidden_size
 
-        if (num_hidden_layers_override is None
-                or num_hidden_layers_override == config.num_hidden_layers):
-            self.need_post_layernorm = True
-        elif num_hidden_layers_override > config.num_hidden_layers:
-            raise ValueError(
-                "num_hidden_layers_override cannot be greater than "
-                "num_hidden_layers")
```
Comment (ywang96) [inline vllm/model_executor/models/siglip.py:452]: We should keep this check no?

**Review by ywang96 — COMMENTED** [review]
Overall LGTM but I left some comments! PTAL

File: vllm/model_executor/models/siglip.py [hunk 2]
```diff
@@ -512,8 +507,8 @@ def __init__(
         )
 
     @property
-    def need_post_layernorm(self):
-        return self.vision_model.need_post_layernorm
+    def need_post_layernorm(self) -> bool:
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/siglip.py]: Fixed

File: vllm/model_executor/models/clip.py [hunk 3]
```diff
@@ -386,9 +395,12 @@ def __init__(self,
             quant_config=quant_config,
             num_hidden_layers_override=num_hidden_layers_override)
 
-    def forward(self, pixel_values: Optional[torch.Tensor] = None):
+    @property
+    def need_post_layernorm(self) -> bool:
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py]: Fixed

File: vllm/model_executor/models/clip.py [hunk 4]
```diff
@@ -355,6 +355,12 @@ def __init__(self,
             quant_config=quant_config,
             num_hidden_layers_override=num_hidden_layers_override)
 
+        if len(self.encoder.layers) == config.num_hidden_layers:
+            self.post_layernorm = nn.LayerNorm(embed_dim,
+                                               eps=config.layer_norm_eps)
+        else:
+            self.post_layernorm = None
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py]: Done

File: vllm/model_executor/models/siglip.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/siglip.py:452]: I've re-added the check but inside the `VisionTransformer` class so the if-else statements stay together.

**Review by ywang96 — APPROVED** [review]

**Comment by litianjian** [comment]
@DarkLight1337  "len(self.encoder.layers) < config.num_hidden_layers"  may not always determine skip the "post_layer_norm" . If model uses the first-last hidden states and skip the "post_layer_norm", int this case "len(self.encoder.layers) == config.num_hidden_layers".

**Comment by DarkLight1337** [comment]
> @DarkLight1337 "len(self.encoder.layers) < config.num_hidden_layers" may not always determine skip the "post_layer_norm" . If model uses the first-last hidden states and skip the "post_layer_norm", int this case "len(self.encoder.layers) == config.num_hidden_layers".

I see what you mean. Is there anything in the HF config that lets you toggle this behavior on or off though?

**Comment by litianjian** [comment]
> off

https://huggingface.co/llava-hf/llava-onevision-qwen2-7b-ov-hf/blob/main/config.json
I am developing the llava-ov models on vLLM. This config triggers this issue.

**Comment by litianjian** [comment]
> > @DarkLight1337 "len(self.encoder.layers) < config.num_hidden_layers" may not always determine skip the "post_layer_norm" . If model uses the first-last hidden states and skip the "post_layer_norm", int this case "len(self.encoder.layers) == config.num_hidden_layers".
> 
> I see what you mean. Is there anything in the HF config that lets you toggle this behavior on or off though?

https://huggingface.co/llava-hf/llava-onevision-qwen2-7b-ov-hf/blob/main/config.json
I am developing the llava-ov models on vLLM. This config triggers this issue.

**Comment by DarkLight1337** [comment]
Looks like SigLIP has `vision_config.use_vision_head` setting. Let me open a PR for this.

**Comment by DarkLight1337** [comment]
Nevermind, I see that the problem actually comes from how the features are extracted in the LLaVA-OneVision model. I think you may have to modify the existing encoders when you develop Llava-OneVision in your PR.

**Comment by ywang96** [comment]
This seems to get unnecessarily complicated than what it should be. Perhaps we can keep track of an internal mapping to see which vision encoders actually use the post layernorm instead of relying on the number of hidden layers as a proxy?

**Comment by DarkLight1337** [comment]
> This seems to get unnecessarily complicated than what it should be. Perhaps we can keep an internal mapping to see which VLM's vision encoders actually use the post layernorm instead of relying on the number of hidden layers as a proxy?

We have several cases:

1. The output state of an intermediate layer is used without layer normalization
2. The output state of the last layer is used without layer normalization
3. The output state of the last layer is used with layer normalization

Just from the config, we can only distinguish between case (1) and case (2, 3). We can add another parameter to the vision encoder model to specify whether to load `post_layernorm`, which would be able to distinguish between case (2) and case (3). Since the vision encoder can be initialized from the VLM, this lets each VLM control whether to use `post_layernorm`.

**Comment by ywang96** [comment]
> > This seems to get unnecessarily complicated than what it should be. Perhaps we can keep an internal mapping to see which VLM's vision encoders actually use the post layernorm instead of relying on the number of hidden layers as a proxy?
> 
> We have several cases:
> 
> 1. The output state of an intermediate layer is used without layer normalization
> 2. The output state of the last layer is used without layer normalization
> 3. The output state of the last layer is used with layer normalization
> 
> Just from the config, we can only distinguish between case (1) and case (2, 3). We could add another parameter to the vision encoder model to specify whether to load `post_layernorm`, which further distinguishes between case (2) and case (3). Since the vision encoder could be initialized from the VLM, this lets each VLM control whether to use `post_layernorm`.

Yea, this is also what I meant by an internal mapping, or as you suggested this can be just an extra parameter in each model's implementation (but the user won't be able to change it from the config). Either way, this is something vLLM "maintains" instead of relying on the config file.

**Comment by litianjian** [comment]
> > > This seems to get unnecessarily complicated than what it should be. Perhaps we can keep an internal mapping to see which VLM's vision encoders actually use the post layernorm instead of relying on the number of hidden layers as a proxy?
> > 
> > 
> > We have several cases:
> > 
> > 1. The output state of an intermediate layer is used without layer normalization
> > 2. The output state of the last layer is used without layer normalization
> > 3. The output state of the last layer is used with layer normalization
> > 
> > Just from the config, we can only distinguish between case (1) and case (2, 3). We could add another parameter to the vision encoder model to specify whether to load `post_layernorm`, which further distinguishes between case (2) and case (3). Since the vision encoder could be initialized from the VLM, this lets each VLM control whether to use `post_layernorm`.
> 
> Yea, this is also what I meant by an internal mapping (to keep track which models need post layernorm, and which not), or as you suggested this can be just an extra parameter in each model's implementation (but the user won't be able to change it from the config). Either way, this is something vLLM "maintains" instead of relying on the config file.

Could you please update me on the progress? This will trigger bugs in LLaVA models with multi videos/images input.

**Comment by DarkLight1337** [comment]
> > > > This seems to get unnecessarily complicated than what it should be. Perhaps we can keep an internal mapping to see which VLM's vision encoders actually use the post layernorm instead of relying on the number of hidden layers as a proxy?
> > > 
> > > 
> > > We have several cases:
> > > 
> > > 1. The output state of an intermediate layer is used without layer normalization
> > > 2. The output state of the last layer is used without layer normalization
> > > 3. The output state of the last layer is used with layer normalization
> > > 
> > > Just from the config, we can only distinguish between case (1) and case (2, 3). We could add another parameter to the vision encoder model to specify whether to load `post_layernorm`, which further distinguishes between case (2) and case (3). Since the vision encoder could be initialized from the VLM, this lets each VLM control whether to use `post_layernorm`.
> > 
> > Yea, this is also what I meant by an internal mapping (to keep track which models need post layernorm, and which not), or as you suggested this can be just an extra parameter in each model's implementation (but the user won't be able to change it from the config). Either way, this is something vLLM "maintains" instead of relying on the config file.
> 
> Could you please update me on the progress? This will trigger bugs in LLaVA models with multi videos/images input. 

This should be fixed already.

**Comment by DarkLight1337** [comment]
Can you open a new issue so it can be easily tracked? Thanks and sorry for the delay, we have been quite swamped lately.

**Comment by litianjian** [comment]
> Can you open a new issue so it can be easily tracked? Thanks and sorry for the delay, we have been quite swamped lately.

Sure. Thanks.

**Comment by litianjian** [comment]
> Can you open a new issue so it can be easily tracked? Thanks and sorry for the delay, we have been quite swamped lately.

FYI. We can use  #9186 to track it.

**Pull request merged**